frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx

```diff
@@ -0,0 +1,118 @@
+import classNames from "classnames";
+import type { FocusEvent, MouseEvent, KeyboardEvent } from "react";
+import { useRef, useState, useMemo } from "react";
+import { t } from "ttag";
+
+import { QueryColumnPicker } from "metabase/common/components/QueryColumnPicker";
+import { Button, Icon, Input, Popover, FocusTrap } from "metabase/ui";
+import * as Lib from "metabase-lib";
+
+import styles from "./ColumnInput.module.css";
+
+type ColumnInputProps = {
+  query: Lib.Query;
+  stageIndex: number;
+  columns: Lib.ColumnMetadata[];
+  label: string;
+  value: Lib.ColumnMetadata | null;
+  onChange: (column: Lib.ColumnMetadata | null) => void;
+};
+
+export function ColumnInput({
+  query,
+  stageIndex,
+  columns,
+  label,
+  value,
+  onChange,
+}: ColumnInputProps) {
+  const columnGroups = useMemo(() => Lib.groupColumns(columns), [columns]);
+
+  const [open, setOpen] = useState(false);
+  const button = useRef<HTMLButtonElement>(null);
+
+  function handleOpen() {
+    setOpen(true);
+  }
+
+  function handleClose() {
+    setOpen(false);
+    button.current?.focus();
+  }
+
+  function handleBlur(evt: FocusEvent<HTMLDivElement>) {
+    if (!evt.currentTarget || !evt.relatedTarget) {
+      return;
+    }
+    if (!evt.currentTarget.contains(evt.relatedTarget as Node)) {
+      setTimeout(() => setOpen(false), 100);
+    }
+  }
+
+  function handleButtonClick(evt: MouseEvent<HTMLButtonElement>) {
+    evt.preventDefault();
+    evt.stopPropagation();
+    setOpen(open => !open);
+  }
+
+  function handleKeyDown(evt: KeyboardEvent<HTMLButtonElement>) {
+    if (evt.key === "Enter") {
+      setOpen(true);
+    }
+  }
+
+  const dropdown = (
+    <FocusTrap active={open}>
+      <div onBlur={handleBlur}>
+        <QueryColumnPicker
+          query={query}
+          stageIndex={stageIndex}
+          columnGroups={columnGroups}
+          onSelect={onChange}
+          onClose={handleClose}
+          checkIsColumnSelected={item => item.column === value}
+          width="100%"
+        />
+      </div>
+    </FocusTrap>
+  );
+
+  const text = useMemo(() => {
+    if (!value) {
+      return t`Select a column...`;
+    }
+    const info = Lib.displayInfo(query, stageIndex, value);
+    return info.longDisplayName;
+  }, [value, query, stageIndex]);
+
+  return (
+    <Input.Wrapper label={label} styles={{ root: { width: "100%" } }}>
+      <Popover
+        opened={open}
+        onClose={handleClose}
+        onOpen={handleOpen}
+        closeOnEscape
+        closeOnClickOutside
+        width="target"
+        returnFocus
+      >
+        <Popover.Target>
+          <Button
```

Comment:
We are emulating a select here with a button.

The initial idea (by Kamil) was to use Mantine's Select component with a custom `dropdownComponent` prop.

Mantine's `Select` does not allow the focus to go to the dropdown component (amongst other issues). It is very hacky and fragile to try and hack the custom dropdown into this component.